<div style="border:solid Chocolate 2px; padding: 40px">

Привет Тим, меня зовут Саша Атаков. Сегодня я проверю твой проект.
<br> Дальнейшее общение будет происходить на "ты" если это не вызывает никаких проблем. Можем работать и на "вы", вдруг тебе неудобно)
<br> Желательно реагировать на каждый мой комментарий ('исправлено', 'не понятно как исправить ошибку', ...)
<br> Пожалуйста, не удаляй комментарии ревьюера, так как они повышают качество повторного ревью.

Комментарии будут в следующих рамках:

<div class="alert alert-block alert-success">
<b>Успех👍:</b> Если все сделано отлично
</div>

<div class="alert alert-block alert-warning">
<b>Совет🤔: </b> Если можно немного улучшить
</div>

<div class="alert alert-block alert-danger">
<b>Ошибка❌:</b> Если требуются исправления. Работа не может быть принята с красными комментариями.
</div>

-------------------

Будет очень хорошо, если ты будешь помечать свои действия следующим образом:
<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> ...
</div>

    
<div class="alert alert-block alert-info">
<b>Изменения:</b> Были внесены следующие изменения ...
</div>

<b><font color='orange'>Вот тебе полезная ссылка, вдруг подчерпнешь что-то новое для себя):</font></b>
* Обзоры по применению криптографии в машинном обучении: https://arxiv.org/pdf/1902.04109.pdf
* Подробнее про аналогичный алгоритм: https://www.cs.cmu.edu/~rjhall/JOS_revised_May_31a.pdf
* Про корреляцию качественных признаков: https://www.statology.org/pearson-correlation-assumptions/

<b> <font color='orange'>Общее впечатление</font></b>
* Большое спасибо за проделанную работу. Видно, что были отлично проработаны знания по матрицам, прекрасно сделано)
* Отлично, что ноутбук хорошо структурирован. Приятно проверять такие работы.
* Над этим проектом нужно будет еще немного поработать, ведь критические ошибки требуют внимания. Однако, изменения не должны занять много времени.
* В работе я оставил несколько советов. Буду рад, если они будут учтены.
* В конце прописал свой комментарий по всему проекту, ознакомься с ним, пожалуйста)
* С радостью отвечу на твои вопросы, касательно выданной тебе темы, если они есть. Жду новую версию проекта.)

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li><li><span><a href="#Общий-вывод-по-проекту" data-toc-modified-id="Общий-вывод-по-проекту-6"><span class="toc-item-num">6&nbsp;&nbsp;</span><b>Общий вывод по проекту</b></a></span></li><li><span><a href="#Общий-вывод-по-проекту" data-toc-modified-id="Общий-вывод-по-проекту-7"><span class="toc-item-num">7&nbsp;&nbsp;</span><b>Общий вывод по проекту</b></a></span></li><li><span><a href="#Общий-вывод-по-проекту-В3" data-toc-modified-id="Общий-вывод-по-проекту-В3-8"><span class="toc-item-num">8&nbsp;&nbsp;</span><b>Общий вывод по проекту В3</b></a></span></li><li><span><a href="#Общий-вывод-по-проекту-В4" data-toc-modified-id="Общий-вывод-по-проекту-В4-9"><span class="toc-item-num">9&nbsp;&nbsp;</span><b>Общий вывод по проекту В4</b></a></span></li></ul></div>

# Защита персональных данных клиентов

Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

<div class="alert alert-block alert-success">
<b>Успех👍:</b> 

Крайне полезно то, что ты написал описание к работе, потому что так можно точно понять, о чем идет речь в данном проекте - молодец)
</div>

In [3]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from sklearn.model_selection import train_test_split

<div class="alert alert-block alert-success">
<b>Успех👍:</b> 

Крайне удобно, когда программист указывает все библиотеки с которыми он работал в самом начале проекта. Молодец!
</div>

## Загрузка данных

Загружаем данные

In [4]:
data = pd.read_csv('/datasets/insurance.csv')

Смотрим информацию о данных

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


In [6]:
data.head()

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0

In [7]:
data.describe()

Пол      Возраст      Зарплата  Члены семьи  Страховые выплаты
count  5000.000000  5000.000000   5000.000000  5000.000000        5000.000000
mean      0.499000    30.952800  39916.360000     1.194200           0.148000
std       0.500049     8.440807   9900.083569     1.091387           0.463183
min       0.000000    18.000000   5300.000000     0.000000           0.000000
25%       0.000000    24.000000  33300.000000     0.000000           0.000000
50%       0.000000    30.000000  40200.000000     1.000000           0.000000
75%       1.000000    37.000000  46600.000000     2.000000           0.000000
max       1.000000    65.000000  79000.000000     6.000000           5.000000

После данной информации по столбцам можно с уверенностью сказать, что в данных не присутствует каких либо аутлайеров либо неверных значений, а значит с ними можно работать дальше, проверим пропуски и дубликаты

<div class="alert alert-block alert-success">
<b>Успех👍:</b> 

Согласен с тобой, вроде ничего лишнего и того, что нужно добавить нет - молодец)
</div>

In [8]:
data.isnull().sum()

Пол                  0
Возраст              0
Зарплата             0
Члены семьи          0
Страховые выплаты    0
dtype: int64

Пропусков нет

In [9]:
data.duplicated().sum()

153

Обнаружено 153 дубликата, их можно удалить без вреда для датасета

In [10]:
data = data.drop_duplicates()
data.duplicated().sum()

0

<div class="alert alert-block alert-warning">
<b>Совет🤔: </b>

Дубликаты не стоит сразу удалять, ведь они могут хранить в себе ценную информацию + в нашем случае, где нет уточнений клиентов (имени и айди), то это могли быть просто совпадения
</div>

Отлично! Теперь можно двигаться дальше

<div class="alert alert-block alert-success">
<b>Успех👍:</b> 

Погнали дальше!
</div>

## Умножение матриц

В этом задании вы можете записывать формулы в *Jupyter Notebook.*

Чтобы записать формулу внутри текста, окружите её символами доллара \\$; если снаружи —  двойными символами \\$\\$. Эти формулы записываются на языке вёрстки *LaTeX.* 

Для примера мы записали формулы линейной регрессии. Можете их скопировать и отредактировать, чтобы решить задачу.

Работать в *LaTeX* необязательно.

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

**Ответ:** Мне кажется качество линейной регресии не должно сильно измениться

**Обоснование:** Я абсолютно не уверен и не очень понимаю, но по моему мнению это связано с векторами и правилами работой линейной регрессии, которая ищет линию, ближающую ко всем остальным линиями, а в данном случае у нас линии просто ддолжны сдвинуться в какую то одну сторону, в связи с чем у нас не должно повредиться качество предсказаний.
Также стоит отдельно отметить то, что при умножении матрицы на какую то другую матрицу, у нас не изменяется расположение векторов, а можно сказать изменяется само пространство в котором эти вектора расположены, и чтобы вернуться к изначальной матрице  нам достаточно всего лишь умножить полученную матрицу на обратную к той, на которую мы умножали нашу изначальную матрицу.

<div class="alert alert-block alert-danger">
<b>Ошибка❌:</b> 
    
ТАкс, ну во-первых будь увереннее в своих прогнозах, ведь ты аналитик, бизнес смотрит на то как ты показываешь результаты, как говоришь о них итп. Поэтому увереннее). А вот главный косяк - доказательство неверно раскрыто + нет формул доказательств - нужно добавить. Можешь начать с этого:
    
Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$
</div>

По условию признаки умножают на обратимую матрицу **P**, т.е. эта матрица является квадратной и не вырожденной. Тогда результат **R** умножения матриц:

$$
R = XP
$$

Тогда формула обучения **W** для новой матрицы **R** примет вид:

$$
W = (R^T R)^{-1} R^T y
$$

Подставив **XP** вместо **R**:

$$
W = ([XP]^T*XP)^{-1}*(XP)^T*y
$$

Вспомним основные свойства, которые нам могут понадобится:

- Сочетательное свойтсво:

    $$
    ABC = (AB)C = A(BC)
    $$
    
- Раскрытие обратимости:

    $$
    (AB)^{-1} = B^{-1}A^{-1}
    $$
    
- Раскрытие транспонирования:

    $$
    (AB)^T = B^TA^T
    $$
    
Далее начинаем работу с формулой:

$$
W' = ([XP]^T*XP)^{-1}*(XP)^T*y
$$

Для начала раскроем скобки

$$
W' = PT^{-1} * P^(-1) * (XT * X) * XT * PT * y 
$$

Теперь сократим 
$$
PT^{-1} * PT = 1
$$
$$
W' = P^(-1) * (XT * X) * XT * y
$$

Заменим
$$
w = (XT * X) * XT * y
$$
Получилось
$$
w' = P^{-1} * w
$$

Значит у нас получается равенство
$$
a' = X'*w' = X * P * P^(-1) * w = Xw = a
$$

<div class="alert alert-block alert-danger">
<b>Ошибка❌ В2:</b> 
    
Ну чтож, сейчас получше, но есть ошибки:
    
- раскрытие скобок при обратимости неверное, в твоем случае надо не забывать про сочетательное свойство: $(ABC)^{-1} = ((AB)C)^{-1} = C^{-1} (AB)^{-1} = C^{-1} B^{-1} A^{-1}$
    
- Также раскрытие скобок при транспонировании у тебя не так выполнено: $(XP)^T = P^T X^T$
    
Поэтому тебе нужно отсюда $W' = ([XP]^T*XP)^{-1}*(XP)^T*y$, пойти по такому сценарию:
    
$$
W' = (P^T X^T XP)^{-1}P^T X^T y
$$
    
$$
W' = ([P^T X^T X] P)^{-1}P^T X^T y
$$
    
$$
W' = P^{-1} (P^T [X^T X])^{-1}P^T X^T y
$$
    
....
</div>

Мне кажется у меня все верно получилось (наверное я просто написать нормально не смог), сейчас попробую исправить, суть то верная :)

$$
W' = ([XP]^T*XP)^{-1}*(XP)^T*y
$$
Раскроем все скобки
$$
W' = (P^{-1}) * (P^T)^{-1} * (X^T)^{-1} * X^{-1} * P^T * X^T * y
$$

Сокращаем
$$
W' = (P^{-1}) * (X^T)^{-1} * X^{-1} * X^T * y
$$

Заменяем
$$
W = (X^T)^{-1} * X^{-1} * X^T * y
$$

W' = (P^{-1}) * W
Далее проверим равенство

$$
a' = X'*W' = X * P * P^{-1} * W = XW = a
$$

<div class="alert alert-block alert-danger">
<b>Ошибка❌ В3:</b> 
    
А теперь у тебя преобразовывание того, что не может сущестовать $W' = (P^{-1}) * (P^T)^{-1} * (X^T)^{-1} * X^{-1} * P^T * X^T * y$. Перепиши так:
    
$$
W = ([P^TX^T]*XP)^{-1}*P^T* X^T * y
$$
        
$$
W = (P^TX^TXP)^{-1}*P^T* X^T * y
$$
        
$$
W = ([P^TX^TX]*P)^{-1}*P^T* X^T * y
$$
        
$$
W = P^{-1}*[P^T(X^T*X)]^{-1}*P^T* X^T * y
$$
        
$$
W = P^{-1}*[X^T*X]^{-1}*[P^T]^{-1}*P^T* X^T * y
$$
    
$$
A = XPP^{-1}*[X^T*X]^{-1}*[P^T]^{-1}*P^T* X^T * y
$$
    
$$
A = XE*[X^T*X]^{-1}*E* X^T * y
$$
    
$$
A = X*[X^T*X]^{-1}* X^T * y
$$
    
$$
A = X*w
$$
    
Что и требовалось доказать
</div>

Раскрываем скобки
$$
W = ([P^TX^T]*XP)^{-1}*P^T* X^T * y
$$
        
$$
W = (P^TX^TXP)^{-1}*P^T* X^T * y
$$
        
$$
W = ([P^TX^TX]*P)^{-1}*P^T* X^T * y
$$
Выносим за скобку P^(-1)        
$$
W = P^{-1}*[P^T(X^T*X)]^{-1}*P^T* X^T * y
$$
        
$$
W = P^{-1}*[X^T*X]^{-1}*[P^T]^{-1}*P^T* X^T * y
$$
Производим замену    
$$
A = XPP^{-1}*[X^T*X]^{-1}*[P^T]^{-1}*P^T* X^T * y
$$
Сокращаем    
$$
A = XE*[X^T*X]^{-1}*E* X^T * y
$$
    
$$
A = X*[X^T*X]^{-1}* X^T * y
$$
Равенство доказано    
$$
A = X*w
$$
    

Я вроде бы примерно понял, но можете чуть чуть подробнее объяснить махинации после замены, я все таки хочу уже хорошо в этом разобраться если я так долго застрял на этом

<div class="alert alert-block alert-success">
<b>Успех👍В4:</b> 

Супер, теперь все верно)
</div>

## Алгоритм преобразования

**Алгоритм**

Я хочу чтобы мой алгоритм строился на рандомных числах и был наименее предсказуемым.
Поэтому выделю следующие ключевые точки.
1. Я умножу нашу матрицу на рандомную матрицу
2. Повторю данную операцию несколько раз
3. Проверю каждую из матриц на обратимость, делать я это буду с помощью обпределителя, так как когда он не равен нулю матрица необратима (Если матрица все же окажется необратимой, то я просто не буду умножать на нее, веростность этого мала, а веростность того, что данные никак не зашифруются также близится к нулю так как при 10 сгенерированный матрицах хотя бы одна в 99,99% точно будет обратима)
4. Чтобы дешифровать данные будет достаточно в обратном порядке умножить нашу полученную матрицу на случайно сгенерированные раннее

**Обоснование**

Это абсолютно понятный код, который раз за разом будет случайно генерироваться и таким образом его будет сложно дешифровать, к тому же данная операция будет повторяться несколько раз, еще больше усложняя процесс. Результаты в ходе данных операций никак не изменятся

<div class="alert alert-block alert-success">
<b>Успех👍:</b> 

Хорошо расписываешь, мне нравится!
</div>

<div class="alert alert-block alert-danger">
<b>Ошибка❌:</b> 
    
Однако я не заметил важный момент - проверить созданную матрицуна обратимость - это важный этап, так как из-за него у нас может возникнуть ошибка. Осуществи проверку либо np.linalg.inv(rand_matrix), либо через определитель матрицы (Если определитель квадратной матрицы P не равен нулю, значит у нее есть обратная)
</div>

<div class="alert alert-block alert-success">
<b>Успех👍 В2:</b> 

Теперь вижу, молодец)
</div>


## Проверка алгоритма

In [11]:
target = data['Страховые выплаты']
features = data.drop('Страховые выплаты', axis = 1)
features_train, features_valid, target_train, target_valid = train_test_split(
    features, target, test_size=0.25, random_state=12345)
model = LinearRegression()
model.fit(features_train, target_train)
predictions = model.predict(features_valid)
print(r2_score(target_valid, predictions))

0.42307727492147584


In [12]:
def transformation(X, y):
    X_trans = X
    for i in range(10):
        trans = np.random.randint (1, 100000, (X.shape[1], X.shape[1]))
        if np.linalg.det(trans) != 0:
            X_trans = X_trans@trans
            features_train, features_valid, target_train, target_valid = train_test_split(
            X_trans, y, test_size=0.25, random_state=12345)
            model = LinearRegression()
            model.fit(features_train, target_train)
            predictions = model.predict(features_valid)
            print(r2_score(target_valid, predictions))
        else:
            continue

<div class="alert alert-block alert-danger">
<b>Ошибка❌ в2:</b> 
    
Смотри, у тебя функция будет работать с обычной матрицой, если у тебя новая, случайная матрица не имеет обратимую, а она у тебя всегда будет обратима, ведь определитель должен быть отличным от нуля для того, чтобы доказать обратимость матрицы. Получается ты проверяешь матрицу неверно и из-за этого все идет неправильно. Сделай проверку не (==),а (!=), так вернее + перекинь весь код после continue в место после   X_trans = X_trans@trans
</div>

<div class="alert alert-block alert-warning">
<b>Совет🤔 В3: </b>

Здесь уже можно было бы цикл убрать for i in range, вместо этого расписать try-except
</div>

Хорошо, сделал так, но мне кажется так получается не совсем то, что я хотел, так мы видим точность только после того, как мы нашу матрицу умножили на одну рандомную, а я хочу умножить ее 10 раз, для большего изменения, не очень понял, почему именно так будет верно, подскажите пожалуйста

С другой стороны можно сделать и так, данный метод тоже имеет свои плюсы

In [13]:
transformation(features, target)

0.4230772749247892
0.4230772746591791
0.4230772798237419
0.4230772736840739
0.4230773024292236
0.42307503475364117
0.4230884364579327
0.4231507940494882
0.4155512934529191
0.41589588209159567


<div class="alert alert-block alert-success">
<b>Успех👍:</b> 

Прекрасно, разницы почти что нет вообще - это замечательный результат, основная цель проекта реализована - молодец
</div>

Результаты получились одинаковые, трансформация работает на ура

<div class="alert alert-block alert-danger">
<b>Ошибка❌:</b> 
    
А вот итогового вывода ко всему проекту я не увидел(. В нем нужно расписать все то, что ты реализуешь по проекту, каких целей достигаешь и общую оценку проекта. Добавь пожалуйста
</div>   

Вывод: 
1) Данные были успешно проверены и обработаны

2) Был разработан алгоритм шифровки данных

3) Алгоритм шифрования данных был успешно проверен на прототипе

4) Была проверена гипотеза о том, что при линейном изменении матрицы результаты регрессии не будут меняться

<div class="alert alert-block alert-warning">
<b>Совет🤔 В2: </b>

Вывод мог бы пополнее расписать, касаясь всего того, что ты сделал по проекту, ведь по проекту ты много что сделал
</div>

## Чек-лист проверки

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные загружены
- [x]  Выполнен шаг 2: получен ответ на вопрос об умножении матриц
    - [x]  Указан правильный вариант ответа
    - [x]  Вариант обоснован
- [x]  Выполнен шаг 3: предложен алгоритм преобразования
    - [x]  Алгоритм описан
    - [x]  Алгоритм обоснован
- [x]  Выполнен шаг 4: алгоритм проверен
    - [x]  Алгоритм реализован
    - [x]  Проведено сравнение качества моделей до и после преобразования

<div style="border:solid Chocolate 2px; padding: 40px">

<h2><b>Общий вывод по проекту</b></h2>

Ты очень постарался, спасибо тебе за такую работу!)

В ходе проекта были выполнены все этапы проекта, главная цель достигнута. Конечно присутствует ряд моментов, которые необходимо рассмотреть, но ничего страшного, для это я здесь). Давай пройдемся по всему что ты сделал, чтобы понять что надо сделать для того, чтобы проект был засчитан:

- в формуле немного не туда свернули, но получили верный ответ - поправь формулу)
- в алгоритме нет ничего касательно проверки на обратимость - впиши ее и реализуй
- итогового вывода нет, а он нам нужен для полной картины проекта - добавь
- Есть советы по проекту, если хочешь и есть время - реализуй их. Если нет, то запомни на будущее)

Работа классная, но нужно доработать! Удачи)

<div style="border:solid Chocolate 2px; padding: 40px">

<h2><b>Общий вывод по проекту</b></h2>

Такс, проект стал лучше, молодец). Однако есть над чем порабоатть:
    
- В формуле есть косяки - поправь
 
- При проверке на обратимость у тебя код немного не так выполняется, как стоило бы - поправь)
    
Удачи!

<div style="border:solid Chocolate 2px; padding: 40px">

<h2><b>Общий вывод по проекту В3</b></h2>

Ну все, тебе осталось лишь переписать формулу, которую мы с тобой вывели и все, проект приму. Давай, быстро справишься)

<div style="border:solid Chocolate 2px; padding: 40px">

<h2><b>Общий вывод по проекту В4</b></h2>

Все, проект выполнен, ты молодец). Удачи в следующих проектах!